src/Orchard.Cms.Web/Modules/Orchard.OpenId/Services/OpenIdService.cs

```diff
@@ -0,0 +1,241 @@
+            {
+                modelState.AddModelError(modelStateKey, T["Invalid url."]);
+                return;
+            }
+            foreach (var uriString in uriStrings.Select(a => (a ?? "").Trim()))
+            {
+                Uri uri;
+                if (!Uri.TryCreate(uriString, UriKind.Absolute, out uri) || ((onlyAllowHttps || uri.Scheme != "http") && uri.Scheme != "https"))
+                {
+                    var message = onlyAllowHttps ? T["Invalid url. Non https urls are only allowed in testing mode."] : T["Invalid url."];
+                    modelState.AddModelError(modelStateKey, T[message]);
+                }
+            }
+        }
+
+        public bool IsValidOpenIdSettings(OpenIdSettings settings)
+        {
+            var modelState = new ModelStateDictionary();
+            return IsValidOpenIdSettings(settings, modelState);
+        }
+
+        public IEnumerable<CertificateInfo> GetAvailableCertificates(bool onlyCertsWithPrivateKey)
+        {
+            foreach (StoreLocation storeLocation in Enum.GetValues(typeof(StoreLocation)))
+            {
+                foreach (StoreName storeName in Enum.GetValues(typeof(StoreName)))
+                {
+                    using (X509Store x509Store = new X509Store(storeName, storeLocation))
+                    {
+                        yield return new CertificateInfo()
+                        {
+                            StoreLocation = storeLocation,
+                            StoreName = storeName
+                        };
+
+                        x509Store.Open(OpenFlags.ReadOnly);
+
+                        X509Certificate2Collection col = x509Store.Certificates;
+                        foreach (var cert in col)
+                        {
+                            if (!cert.Archived && (!onlyCertsWithPrivateKey || (onlyCertsWithPrivateKey && cert.HasPrivateKey)))
+                            {
+                                yield return new CertificateInfo()
+                                {
+                                    StoreLocation = storeLocation,
+                                    StoreName = storeName,
+                                    FriendlyName = cert.FriendlyName,
+                                    Issuer = cert.Issuer,
+                                    Subject = cert.Subject,
+                                    NotBefore = cert.NotBefore,
+                                    NotAfter = cert.NotAfter,
+                                    ThumbPrint = cert.Thumbprint,
+                                    HasPrivateKey = cert.HasPrivateKey,
+                                    Archived = cert.Archived
+                                };
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public CertificateInfo GetCertificate(StoreLocation storeLocation, StoreName storeName, string certThumbprint)
```

Comment:
@Jetski5822 you can't do that: `CertificateInfo` is a class. `Nullable<T>`/`T?` only works with value types (structs).